### PR TITLE
box: fix _func id overflow

### PR DESCRIPTION
**Pull request by LevKats**
Each time one used the `box.schema.func.create`, the primary key of
the resulted function increased because of the
`_func:auto_increment`. Now the new key is chosen by the
`generate_func_id` in Lua and `OP_GenFuncid` SQL opcode, both implemented
using the new `box_generate_func_id` function that shares logic with
`box_generate_space_id`. Hence we are scanning the entire range of
possible ids `[0, BOX_FUNCTION_MAX)` before we are returning an
error.

Fixes https://github.com/tarantool/tarantool/issues/11849
Fixes https://github.com/tarantool/tarantool/issues/11851
NO_DOC=bugfix

### CODE REVIEW

File: changelogs/unreleased/box-func-id-overflow-fix.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## bugfix/box
+
+* Fixed a bug when function IDs grew monotonously, resulting in an overflow
+  after some amount of function modifications, even if the total number of
+  functions is constant (gh-11849, gh-11851).
```
Comment (lenkis) [inline changelogs/unreleased/box-func-id-overflow-fix.md]:
```suggestion
  functions was constant (gh-11849, gh-11851).
```

File: src/box/sql/build.c [hunk 2]
```diff
@@ -1166,8 +1166,8 @@ vdbe_emit_create_function(struct Parse *parser, int reg_id, const char *name,
 					  BOX_FUNC_FIELD_NAME, name_reg, 1,
 					  ER_FUNCTION_EXISTS, name, false,
 					  OP_NoConflict);
-	sqlVdbeAddOp3(v, OP_NextSystemSpaceId, BOX_FUNC_ID,
-		      regs + BOX_FUNC_FIELD_ID, BOX_FUNC_FIELD_ID);
+	sqlVdbeAddOp4(v, OP_GenFuncid, regs + BOX_FUNC_FIELD_ID, 0, 0,
```
Comment (ImeevMA) [inline src/box/sql/build.c]: Can't we use `name_reg` instead of copying `name`?

File: src/box/sql/vdbe.c [hunk 3]
```diff
@@ -2945,13 +2945,13 @@ case OP_Sequence: {           /* out2 */
...
- * Place the next value of the primary key of the _sequence or _func space into
- * register P2. P1 is the system space identifier. P3 is fieldno of primary key.
+ * Place the next value of the primary key of the _sequence space into register
+ * P2. P1 is the system space identifier. P3 is fieldno of primary key.
  */
 case OP_NextSystemSpaceId: {
 	assert(pOp->p1 >= 0 && pOp->p3 >= 0);
 	uint32_t space_id = pOp->p1;
-	assert(space_id == BOX_SEQUENCE_ID || space_id == BOX_FUNC_ID);
+	assert(space_id == BOX_SEQUENCE_ID);
```
Comment (ImeevMA) [inline src/box/sql/vdbe.c]: I believe that using a generic opcode for a single system space is pointless. Please refactor the opcode so that it generates an identifier only for `_sequence`.

File: src/box/sql/vdbe.c [hunk 4]
```diff
@@ -4370,6 +4370,24 @@ case OP_Init: {          /* jump */
 	goto jump_to_p2;
 }
 
+/* Opcode: GenFuncid P1 * * P4 *
```
Comment (ImeevMA) [inline src/box/sql/vdbe.c]: I think it should be `GenFuncId`, right? Also, I think it would be better to name this opcode `OP_NextFuncId`.

File: src/box/sql/vdbe.c [hunk 5]
```diff
@@ -4370,6 +4370,24 @@ case OP_Init: {          /* jump */
...
+ * Synopsis: r[P1]=new func ID
+ *
+ * Generate unique id for a new function and store it in register specified by
+ * first operand. It is system opcode and must be used only during DDL routine.
```
Comment (ImeevMA) [inline src/box/sql/vdbe.c]: I don't understand why you're saying it should only be used for DDL routing. Is there a specific reason for that?

File: changelogs/unreleased/box-func-id-overflow-fix.md (hunk 1 above)
Comment (LevKats) [reply changelogs/unreleased/box-func-id-overflow-fix.md]: Fixed

File: src/box/sql/build.c (hunk 2 above)
Comment (LevKats) [reply src/box/sql/build.c]: Fixed

File: src/box/sql/vdbe.c (hunk 3 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: Now it is `OP_NextSequenceId`

File: src/box/sql/vdbe.c (hunk 5 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: No specific reason, I guess. I started with `OP_Genspaceid` and since the logic is the same for spaces and for functions, I took that comment too. Fixed.

File: src/box/sql/vdbe.c (hunk 4 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: Fixed

File: src/box/sql/vdbe.c [hunk 6]
```diff
@@ -2942,30 +2942,26 @@ case OP_Sequence: {           /* out2 */
 	break;
 }
 
-/* Opcode: NextSystemSpaceId P1 P2 P3 * *
- * Synopsis: r[P2]=New ID of space P1.
+/* Opcode: NextSequenceId P1 * * * *
+ * Synopsis: r[P1]=New sequence ID.
```
Comment (ImeevMA) [inline src/box/sql/vdbe.c]: Please use P2 for the new identifier. Here and for the new opcode. Also, please use the following method for comments:
```
/**
 * Opcode: ...
```

**Review by ImeevMA — APPROVED** [review]
Thank you for the fixes! I have just one comment for SQL part.

I have a few comments on the non-SQL part:
1. I still think that there is no need to complicate the code due to function ID generation.
2. Name of the test-file usually starts with `gh-...` for non-general tests. Also, you can just add the test to `box-luatest/func_test.lua`.

But I leave this part to @sergepetrenko

File: src/box/sql/vdbe.c (hunk 6 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: Fixed, I guess

File: src/box/box.cc [hunk 7]
```diff
@@ -6555,25 +6555,96 @@ box_read_ffi_enable(void)
...
+static int
+box_generate_unique_id(uint32_t space_id, uint32_t index_id,
+		       uint32_t id_range_begin, uint32_t id_range_end,
+		       bool (*is_id_occupied)(uint32_t id),
+		       uint32_t *new_id)
+{
+	assert(new_id != NULL);
+
+	char key_buf[16];
+	char *key_end = key_buf;
```
Comment (sergepetrenko) [inline src/box/box.cc:6634]: Please move factoring out `box_generate_unique_id` in a separate preparatory commit.

File: src/box/box.cc [hunk 8]
```diff
@@ -6555,25 +6555,96 @@ box_read_ffi_enable(void)
...
+ * \param id_range_end end of the range (exclusive)
+ * \retval key if found
+ * \retval value > id_range_end otherwise
+ */
+static uint32_t
+box_find_next_unused_id(uint32_t cur_id, uint32_t id_range_end,
+			bool (*is_id_occupied)(uint32_t id))
+{
+	for (; cur_id < id_range_end; cur_id++)
+		if (!is_id_occupied(cur_id))
```
Comment (sergepetrenko) [inline src/box/box.cc]: Better introduce `space_find_nex_unused_id` and `func_find_next_unused_id` and pass them by pointer to `box_generate_unique_id`.

File: src/box/box.cc [hunk 9]
```diff
@@ -6555,25 +6555,96 @@ box_read_ffi_enable(void)
...
 		(uint32_t)BOX_SPACE_ID_TEMPORARY_MIN :
 		(uint32_t)BOX_SPACE_MAX + 1;
-	char key_buf[16];
-	char *key_end = key_buf;
-	key_end = mp_encode_array(key_end, 1);
-	key_end = mp_encode_uint(key_end, id_range_end);
 	struct credentials *orig_credentials = effective_user();
 	fiber_set_user(fiber(), &admin_credentials);
 	auto guard = make_scoped_guard([=] {
 		fiber_set_user(fiber(), orig_credentials);
```
Comment (sergepetrenko) [inline src/box/box.cc:6605]: I think admin credentials should be used when searching for free func id as well. Let's move this code there.
Otherwise your code will fail finding a spare func id when the user doesn't have read rights for space _func, it seems?

File: src/box/box.cc (hunk 7 above)
Comment (LevKats) [reply src/box/box.cc:6634]: Fixed, I guess

File: src/box/box.cc (hunk 8 above)
Comment (LevKats) [reply src/box/box.cc]: It is `func_cache_find_next_unused_id` (introduced) and `space_cache_find_next_unused_id` (refactored).

File: src/box/box.cc (hunk 9 above)
Comment (LevKats) [reply src/box/box.cc:6605]: Looks like we should do the opposite. The [documentation](https://www.tarantool.io/en/doc/latest/admin/access_control/#privileges) says:
> Note that some privileges might require read and write access to certain system spaces. For example, the create permission granted to the space object requires read and write permissions to the _space system space. Similarly, granting the ability to create functions requires read and write access to the _func space.

Similarly, [wiki](https://github.com/tarantool/tarantool/wiki/Privileges-and-Access-Control-explained#all-possible-combinations-of-object-type-and-permission) says:
> "create" - e.g. allow [box.schema.space.create](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/user_create/) (needs "read,write" for space "_space")

But we may want to keep admin privileges for spaces in 2.11 though.

File: src/box/box.cc (hunk 8 above)
Comment (sergepetrenko) [reply src/box/box.cc]: Thanks!

File: src/box/box.cc (hunk 9 above)
Comment (sergepetrenko) [reply src/box/box.cc:6605]: Hm, ok, seems we can be more strict here indeed.

File: src/box/box.cc [hunk 10]
```diff
@@ -6555,25 +6555,68 @@ box_read_ffi_enable(void)
...
+ * \param id_range_begin min index in the search range (inclusive)
+ * \param id_range_end max index in the search range (exclusive)
+ * \param find_next_unused_id pointer to the function that finds unused id in
+ * the space in the range [cur_id, id_range_end)
+ * \new_id pointer to result
+ * \retval -1 on error (and set new_id to id_range_end)
+ * \retval 0 on success
+ */
+static int
+box_generate_unique_id(uint32_t space_id, uint32_t index_id,
```
Comment (drewdzzz) [inline src/box/box.cc:6569]: Why do we need a forward declaration here?

File: src/box/box.cc [hunk 11]
```diff
@@ -6555,25 +6555,68 @@ box_read_ffi_enable(void)
...
 	fiber_set_user(fiber(), &admin_credentials);
 	auto guard = make_scoped_guard([=] {
 		fiber_set_user(fiber(), orig_credentials);
 	});
-	box_iterator_t *it = box_index_iterator(BOX_SPACE_ID, 0, ITER_LT,
+	int ret = box_generate_unique_id(BOX_SPACE_ID, 0, id_range_begin,
+					 id_range_end,
+					 space_cache_find_next_unused_id,
+					 new_space_id);
+	if (*new_space_id == id_range_end) {
```
Comment (drewdzzz) [inline src/box/box.cc:6612]: It seems that in the case of OOM new_space_id is not initialized and can be accidentally equal to `id_range_end`.
It's very difficult to cover it with a test case, so let's just set `new_id` to zero in the beginning of `box_generate_unique_id`. Also we should leave a note about it in its description.

File: test/box/access.result [hunk 12]
```diff
@@ -59,15 +59,15 @@ session.su('test')
 -- in future we may  introduce a separate privilege
 box.schema.space.create('test')
 ---
-- error: Write access to space '_space' is denied for user 'test'
+- error: Read access to space '_space' is denied for user 'test'
```
Comment (drewdzzz) [inline test/box/access.result]: AFAICS, there are no tests checking write access to space `_space` after this change. Let's add such one.

File: src/box/box.cc
Comment (drewdzzz) [inline src/box/box.cc:1]: > This change allows us to further unify the id generation logic between `_space` and `_func`.

I don't see how it allows to unify the logic. These ids are still generated by different functions.

File: src/box/box.cc [hunk 13]
```diff
@@ -6575,6 +6575,21 @@ box_generate_unique_id(uint32_t space_id, uint32_t index_id,
...
 		       uint32_t *new_id);
 
+int
+box_generate_func_id(uint32_t *new_func_id, const char *func_name)
+{
+	uint32_t id_range_end = BOX_FUNCTION_MAX + 1;
+	int ret = box_generate_unique_id(BOX_FUNC_ID, 0, 0, id_range_end,
+					 func_cache_find_next_unused_id,
+					 new_func_id);
+	if (*new_func_id == BOX_FUNCTION_MAX + 1) {
```
Comment (drewdzzz) [inline src/box/box.cc]: Ditto.

File: src/box/sql/vdbe.c [hunk 14]
```diff
@@ -4370,6 +4367,23 @@ case OP_Init: {          /* jump */
 	goto jump_to_p2;
 }
 
+/**
+ * Opcode: NextFuncId P1 P2 * * *
+ * Synopsis: r[P2]=new func ID
+ *
+ * Generate unique id for a new function and store it into register P2.
+ */
+case OP_NextFuncId: {
```
Comment (drewdzzz) [inline src/box/sql/vdbe.c]: Why not `GenFuncId` for consistency with `GenSpaceId`?
Also, it't basically not a `next` method anymore because it can generate an id that is lower then the previous one.

File: src/box/sql/vdbe.c [hunk 15]
```diff
@@ -4370,6 +4367,23 @@ case OP_Init: {          /* jump */
...
+	assert(pOp->p2 > 0);
+	const char *func_name = aMem[pOp->p1].z;
```
Comment (drewdzzz) [inline src/box/sql/vdbe.c:4378]: Shouldn't we leave a note about this argument in the function's description?

File: test/wal_off/func_max.result [hunk 16]
```diff
@@ -42,11 +42,11 @@ test_run:cmd("setopt delimiter ''");
 ...
 func_limit()
 ---
-- error: 'Failed to create function ''func31936'': function id is too big'
+- error: 'Failed to create function ''func32000'': Max function count is reached'
```
Comment (drewdzzz) [inline test/wal_off/func_max.result]: Hm, why did the amount of functions has changed? xD

File: src/box/box.cc (hunk 13 above)
Comment (drewdzzz) [inline src/box/box.cc]: Why not compare to `id_range_end` here?

File: test/wal_off/func_max.result (hunk 16 above)
Comment (drewdzzz) [inline test/wal_off/func_max.result]: AFAICS, function with `BOX_FUNCTION_MAX` id is still allowed to be created. Shouldn't it be `funct32001` then?

File: test/box-luatest/box_func_no_id_overflow_test.lua [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+            box.schema.func.create("first")
+            box.schema.func.drop("second")
+            box.schema.func.create("second")
+        end
+    end)
+end
+
+g.after_all(function(g)
+    g.server:drop()
+end)
```
Comment (drewdzzz) [inline test/box-luatest/box_func_no_id_overflow_test.lua]: Nit: usually we put `after_all` hook before all test cases, right after `before_all`, so that all hooks are in one place.

File: src/box/box.h [hunk 18]
```diff
@@ -905,6 +905,10 @@ boxk(int type, uint32_t space_id, const char *format, ...);
 int
 box_generate_space_id(uint32_t *new_space_id, bool is_temporary);
 
+/** Generate unique if for a function. */
```
Comment (drewdzzz) [inline src/box/box.h]: Typo: `if` -> `id`.

File: test/box-luatest/box_func_no_id_overflow_test.lua [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+local t = require('luatest')
+
+local g = t.group()
+
+g.before_all(function(g)
+    g.server = server:new()
+    g.server:start()
+end)
+
+g.test_function_create_drop_multiple = function(g)
```
Comment (drewdzzz) [inline test/box-luatest/box_func_no_id_overflow_test.lua]: Is new SQL logic covered in other tests?

File: src/box/box.cc [hunk 20]
```diff
@@ -6555,25 +6555,68 @@ box_read_ffi_enable(void)
...
+ * Generate a unique integer id in a given index in a given space within a
+ * given range, or return -1 if it is not possible.
+ * \param space_id id of the space
+ * \param index_id id of the index in the space
+ * \param id_range_begin min index in the search range (inclusive)
+ * \param id_range_end max index in the search range (exclusive)
+ * \param find_next_unused_id pointer to the function that finds unused id in
+ * the space in the range [cur_id, id_range_end)
+ * \new_id pointer to result
+ * \retval -1 on error (and set new_id to id_range_end)
```
Comment (drewdzzz) [inline src/box/box.cc]: > set new_id to id_range_end

It's not on OOM (and it shouldn't be).

Also, what's about diag?

File: src/box/box.cc [hunk 21]
```diff
@@ -6587,20 +6630,15 @@ box_generate_space_id(uint32_t *new_space_id, bool is_temporary)
...
-		 * memory, and its pretty hard to test it, so let's just panic
-		 * if we've run out of ids.
-		 */
-		if (*new_space_id >= id_range_end)
-			panic("Space id limit is reached");
+	if (*new_id >= id_range_end) {
+		*new_id = find_next_unused_id(id_range_begin, id_range_end);
+		if (*new_id >= id_range_end) {
+			*new_id = id_range_end;
+			return -1;
```
Comment (drewdzzz) [inline src/box/box.cc]: It's not convenient API when sometimes `diag` is set on error and sometimes is not.
Probably, we should return 0 on overflow and set `new_id` to `id_range_end` (so that overflow is not treated as an error at all)

File: test/box-luatest/box_func_no_id_overflow_test.lua [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+    g.server:exec(function()
+        local BOX_FUNCTION_MAX = 32000
+        box.schema.func.create("first")
```
Comment (drewdzzz) [inline test/box-luatest/box_func_no_id_overflow_test.lua]: Nit: we could wrap it into a transaction so the test would run faster.

File: src/box/box.cc (hunk 10 above)
Comment (drewdzzz) [reply src/box/box.cc:6569]: Discussed f2f - decided to leave it as is to make the diff more clear.

File: test/box-luatest/box_func_no_id_overflow_test.lua
Comment (drewdzzz) [inline test/box-luatest/box_func_no_id_overflow_test.lua:1]: Nit: let's add `gh_<issue1>_<issue2>_` prefix to the file name for easy search.

File: src/box/sql/vdbe.c (hunk 14 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: Sorry, missed that comment. It was named this way [here](https://github.com/tarantool/tarantool/pull/12004#discussion_r2477513721). Maybe it is not meant to be consistent with spaces since it is not for DDL operations, but used internally in SQL requests processing. @ImeevMA , what do you think?

File: test/wal_off/func_max.result (hunk 16 above)
Comment (LevKats) [reply test/wal_off/func_max.result]: Initially right after `box.cfg` call we already have 2 entries in the `_func` space: `box.schema.user.info` with id 0 and `LUA` with id 65. Since ids were monotonically increasing before the patch, we used to start inserting new spaces with id 66, but now we can use id=0 (by the way, should we start with 1?), id=2, etc. We may want to reserve some range for system functions though. What do you think?

File: src/box/sql/vdbe.c (hunk 14 above)
Comment (ImeevMA) [reply src/box/sql/vdbe.c]: I agree that it makes sense to rename it to `GenFuncId`, although both options, in my opinion, are quite good. I seem to have forgotten about `GenSpaceId` during the first review, sorry for the misleading.

File: src/box/box.cc (hunk 11 above)
Comment (LevKats) [reply src/box/box.cc:6612]: I decided that the function would return -1 only in case of an error and id overflow is not considered as an error, as you suggested. Fixed, I guess

File: test/box/access.result (hunk 12 above)
Comment (LevKats) [reply test/box/access.result]: Fixed both for spaces and functions.

File: src/box/box.cc
Comment (LevKats) [reply src/box/box.cc:1]: I guess the right word is "consistent". Fixed the commit message.

File: src/box/box.cc (hunk 13 above)
Comment (LevKats) [reply src/box/box.cc]: Fixed.

File: src/box/sql/vdbe.c (hunk 14 above)
Comment (LevKats) [reply src/box/sql/vdbe.c]: Decided to use `GenFuncId `

File: src/box/sql/vdbe.c (hunk 15 above)
Comment (LevKats) [reply src/box/sql/vdbe.c:4378]: Added to Synopsis

File: src/box/box.cc (hunk 13 above)
Comment (LevKats) [reply src/box/box.cc]: Fixed.

File: test/wal_off/func_max.result (hunk 16 above)
Comment (LevKats) [reply test/wal_off/func_max.result]: See the [comment](https://github.com/tarantool/tarantool/pull/12004#discussion_r2564736695). Two were already allocated. It is `func31999` now though because I decided to start func ids with 1.

File: test/box-luatest/box_func_no_id_overflow_test.lua (hunk 17 above)
Comment (LevKats) [reply test/box-luatest/box_func_no_id_overflow_test.lua]: Fixed.

File: src/box/box.h (hunk 18 above)
Comment (LevKats) [reply src/box/box.h]: Fixed.

File: test/box-luatest/box_func_no_id_overflow_test.lua (hunk 19 above)
Comment (LevKats) [reply test/box-luatest/box_func_no_id_overflow_test.lua]: It is now. Fixed, I hope.

File: src/box/box.cc (hunk 20 above)
Comment (LevKats) [reply src/box/box.cc]: Fixed.

File: src/box/box.cc (hunk 21 above)
Comment (LevKats) [reply src/box/box.cc]: Fixed. See one of the comments above.

File: test/box-luatest/box_func_no_id_overflow_test.lua
Comment (LevKats) [reply test/box-luatest/box_func_no_id_overflow_test.lua:1]: Fixed.

File: test/box-luatest/box_func_no_id_overflow_test.lua (hunk 22 above)
Comment (LevKats) [reply test/box-luatest/box_func_no_id_overflow_test.lua]: I didn't succeed here. If I wrap the whole loop with a transaction it results in "fiber slice is exceeded" and if I yield somewhere, it fails the transaction. Should I create a transaction for the loop body?

File: test/wal_off/func_max.result (hunk 16 above)
Comment (drewdzzz) [reply test/wal_off/func_max.result]: > We may want to reserve some range for system functions though.

That's a good idea! Though, I can't imagine why we might want to create a system entry in `_func`. Do you know why we need these `box.schema.user.info` and `LUA` functions there?

File: test/box-luatest/box_func_no_id_overflow_test.lua (hunk 22 above)
Comment (drewdzzz) [reply test/box-luatest/box_func_no_id_overflow_test.lua]: Use `fiber.set_slice(10)` to increase current slice to 10 seconds.
You can use another slice, I just guess that 10 is enough. It should be large enough - the test will be flaky on slow runners otherwise.

File: test/wal_off/func_max.result (hunk 16 above)
Comment (drewdzzz) [reply test/wal_off/func_max.result]: Discussed f2f. Let's leave it as is for the sake of simplicity and do not mess with upgrade without a good reason.

File: src/box/lua/misc.cc [hunk 23]
```diff
@@ -374,6 +374,21 @@ lbox_generate_space_id(lua_State *L)
 	return 1;
 }
 
+/**
+ * Takes function name and generates unique id for a function.
+ * Raises Lua error if all ids are already occupied.
+ */
+static int
+lbox_generate_func_id(lua_State *L)
+{	size_t func_name_len;
```
Comment (drewdzzz) [inline src/box/lua/misc.cc]: Code on the same line with brace.
Surprised that `checkpatch` didn't notice.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+                box.schema.func.create('func'..i)
+                i = i + 1
+            end
+        end)
+        local error_msg = box.error.last().message
+        box.schema.func.drop('func1')
+        box.schema.func.drop('func2')
+        -- The following query creates 2 functions. If it fails, it changes
+        -- the content of box.error.last().
+        box.execute([[
```
Comment (drewdzzz) [inline test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: `box.execute` returns `ok, err` values. Let's simply check `ok`.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+
+g.test_sql_execute_query_func_about_to_overflow = function(g)
+    g.server:exec(function()
+        t.assert_error(function()
+            local i = 1
+            while true do
+                box.schema.func.create('func'..i)
+                i = i + 1
+            end
+        end)
```
Comment (drewdzzz) [inline test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: We should use transaction here as well.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+        box.session.su('admin')
+        box.schema.user.grant('testuser', 'write', 'space', '_func')
+        box.schema.user.revoke('testuser', 'read', 'space', '_func')
+        box.session.su('testuser')
+        t.assert_error(function()
+            box.schema.func.create('test2')
+        end)
+    end)
+end
```
Comment (drewdzzz) [inline test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: These tests don't belong here - we have `access` diff test - let's put them there.

Also, `_space` access rights checks belong to the previous commit.

File: src/box/lua/misc.cc (hunk 23 above)
Comment (LevKats) [reply src/box/lua/misc.cc]: Looks like a valid Horstmann C code style. Fixed :)

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 24 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: Now we check that the `err` is `nil`. Fixed.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 25 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: Done.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 26 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: Moved to `access.test` in the second commit in case we do insist in this behaviour.

File: test/box-luatest/box_func_no_id_overflow_test.lua (hunk 22 above)
Comment (LevKats) [reply test/box-luatest/box_func_no_id_overflow_test.lua]: Fixed.

File: src/box/box.cc
Comment (LevKats) [reply src/box/box.cc:1]: The commit was dropped.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 26 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: The commit was dropped.

File: src/box/box.cc [hunk 27]
```diff
@@ -6583,20 +6630,13 @@ box_generate_space_id(uint32_t *new_space_id, bool is_temporary)
...
-		 * This situation cannot happen in real world with limited
-		 * memory, and its pretty hard to test it, so let's just panic
-		 * if we've run out of ids.
-		 */
-		if (*new_space_id >= id_range_end)
-			panic("Space id limit is reached");
+	if (*new_id >= id_range_end) {
+		*new_id = find_next_unused_id(id_range_begin, id_range_end);
+		if (*new_id >= id_range_end)
+			*new_id = id_range_end;
```
Comment (drewdzzz) [inline src/box/box.cc]: Nit: since `find_next_unused_id` returns `id_range_end` on overflow anyway, it can be rewritten to
```c
if (*new_id == id_range_end)
		*new_id = find_next_unused_id(id_range_begin, id_range_end);
```

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+        -- Let's drop 2 two non-consecutive functions.
+        box.schema.func.drop('func3')
+        box.schema.func.drop('func16')
+        -- The following query creates 2 functions.
+        local _, err = box.execute([[
+            CREATE TABLE t(
+                i INT PRIMARY KEY CHECK (i > 10) DEFAULT (RANDOM())
+            );
+        ]])
+        t.assert_equals(err)
```
Comment (drewdzzz) [inline test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: Nit: `t.assert_equals(err, nil)`.

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+    end)
+end
+
+g.test_sql_execute_query_func_about_to_overflow = function(g)
+    g.server:exec(function()
+        local fiber = require("fiber")
+        fiber.set_slice(60)
+        box.begin()
+        local i = 1
+        while true do
```
Comment (drewdzzz) [inline test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua:39]: Nit: let's use `t.assert_error_msg_content_equals` instead.

File: src/box/box.cc (hunk 27 above)
Comment (LevKats) [reply src/box/box.cc]: Fixed

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 28 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua]: Fixed

File: test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua (hunk 29 above)
Comment (LevKats) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua:39]: Done

Comment (drewdzzz) [reply test/box-luatest/gh_11849_11851_box_func_no_id_overflow_test.lua:39]: I meant we should put the whole cycle into `t.assert_error_msg_content_equals` :)
Anyway, `full-ci` has passed and everything else is OK, so let's leave it as is - it's not critical.

**Comment by drewdzzz** [comment]
I'm OK with the latest fixes, all the comments are resolved and `full-ci` has passed, so I beleive the PR is ready to be merged.

**Review by sergepetrenko — APPROVED** [review]
Lev, thanks for the patch!
